examples/5_secret_storage.py

```diff
@@ -0,0 +1,419 @@
+        self.kvstore: KVStoreClient = bot.kvstore
+
+    async def load_secret(self, team, namespace) -> bytes:
+        if team not in self.secrets or namespace not in self.secrets[team]:
+            secret = secrets.token_bytes(self.SECRET_NUM_BYTES)
+            try:
+                # we don't expect self.SECRET_KEY's revision > 0
+                await self.kvstore.put(
+                    team, namespace, self.SECRET_KEY, bytes_to_str(secret), revision=1
+                )
+            except RevisionError:
+                res: keybase1.KVGetResult = await self.kvstore.get(
+                    team, namespace, self.SECRET_KEY
+                )
+                secret = str_to_bytes(res.entry_value)
+            if team not in self.secrets:
+                self.secrets[team] = {}
+            self.secrets[team][namespace] = secret
+        return self.secrets[team][namespace]
+
+    async def hmac_key(self, team, namespace, entry_key) -> str:
+        secret = await self.load_secret(team, namespace)
+        return hmac.new(secret, entry_key.encode("utf-8")).hexdigest()
+
+    async def put(
+        self,
+        team: str,
+        namespace: str,
+        entry_key: str,
+        entry_value: Dict[str, str],
+        revision: Union[int, None] = None,
+    ) -> keybase1.KVPutResult:
+        entry_value[SecretKeyKVStoreClient.KEY_KEY] = entry_key
+        h = await self.hmac_key(team, namespace, entry_key)
+        res = await self.kvstore.put(
+            team, namespace, h, json.dumps(entry_value), revision
+        )
+        res.entry_key = entry_key
+        return res
+
+    async def delete(
+        self,
+        team: str,
+        namespace: str,
+        entry_key: str,
+        revision: Union[int, None] = None,
+    ) -> keybase1.KVDeleteEntryResult:
+        h = await self.hmac_key(team, namespace, entry_key)
+        res = await self.kvstore.delete(team, namespace, h, revision)
+        res.entry_key = entry_key
+        return res
+
+    async def get(
+        self, team: str, namespace: str, entry_key: str
+    ) -> keybase1.KVGetResult:
+        h = await self.hmac_key(team, namespace, entry_key)
+        res = await self.kvstore.get(team, namespace, h)
+        res.entry_key = entry_key
+        return res
+
+    async def list_entrykeys(
+        self, team: str, namespace: str
+    ) -> keybase1.KVListEntryResult:
+        res = await self.kvstore.list_entrykeys(team, namespace)
+        if res.entry_keys:
+            for e in res.entry_keys:
+                if not e.entry_key.startswith("_"):
+                    get_res = await self.kvstore.get(team, namespace, e.entry_key)
+                    e.entry_key = json.loads(get_res.entry_value)[self.KEY_KEY]
+        return res
+
+
+class RentalClient:
+    """
+        Wraps a CachedKVStoreClient to expose methods to handle tool rentals.
+    """
+
+    NAMESPACE = "rental"
+
+    def __init__(self, kvstore: CachedKVStoreClient):
```

Comment:
i don't know much about duck typing with python's type system, but this would be a case for it. it looks to me like this could take a CachedKVStoreClient or a SecretClient or a regular ol' client. 

oh actually, what i suggested before was to pass in a Bot to this thing. yeah. in which case the ducktyping would be any object that has a `kvstore` property. 

this isn't important.

Reply:
ahh. i see this is calling additional methods on the kvstore that don't exist on the basic one. ok. nevermind.

Reply:
ok - i think i'd argue there's not really a compelling reason to expose `get_cache` to the outside world. the only reason something outside of the CachedKVStoreClient would want to call it is for performance. and that's not really the point of this thing, it's for concurrency. plus it adds branching logic in this client that's just not really necessary. so yeah. i think i'd argue:
1. the cached kvstore client shouldn't expose `get_cached` to the outside world
2. the rental client should always just call `get` (and now it never needs to worry about getting outdated data)
3. the rental client can be instantiated with anything kvstore-like. even if that's not a duck-typing exercise worth doing